### PR TITLE
Megatron Encoder Decoder models with RPE and PP > 2

### PR DESCRIPTION
**Pull request by MaximumEntropy**
Signed-off-by: MaximumEntropy <sandeep.subramanian.1@umontreal.ca>

# What does this PR do ?

Adds support for Enc-Dec models with RPE and PP > 2

**Collection**: NLP

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* model.encoder.position_embedding_type="relative"
* model.decoder.position_embedding_type="relative"
* model.pipeline_model_parallel_size=4
* model.pipeline_model_parallel_split_rank=2

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py [hunk 1]
```diff
@@ -369,35 +369,88 @@ def allreduce_gradients(self):
...
+                torch.distributed.all_reduce(
+                    grad, group=parallel_state.get_encoder_relative_position_embedding_group()
+                )
+
+            # For split rank == pipeline_world_size - 1, we have only one decoder rank and so we don't need to allreduce.
+            if (
+                self.cfg.decoder.get('position_embedding_type') == 'relative'
+                and parallel_state.is_rank_in_decoder_relative_position_embedding_group()
+                and parallel_state.get_pipeline_model_parallel_split_rank()
+                < parallel_state.get_pipeline_model_parallel_world_size() - 1
```
Comment (khcs) [inline nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:431]: We decided these lines are unnecessary and reduce the code readability so will remove.

File: nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py [hunk 2]
```diff
@@ -209,26 +211,28 @@ def __init__(
 
                 self._decoder_embedding_key = "decoder_embedding"
                 # TODO (sandeepsub): When implementing RPE for PP > 2, this should not be inside `pre_process`. It should exist on all ranks and be synchronized manually.
```
Comment (khcs) [inline nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py:213]: This #TODO can be removed.

**Review by khcs — COMMENTED** [review]
Outside of those where I left comments, we made a few changes to the files to make it work.

**Comment by khcs** [comment]
Currently working with PP>2.
Continuing testing with different configs like bf16, O2, PP>4, different encoder/decoder RPE settings, etc.

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py [hunk 3]
```diff
@@ -823,12 +890,12 @@ def on_pretrain_routine_start(self) -> None:
     def setup_validation_data(self, cfg):
         if hasattr(self, '_validation_ds'):
             consumed_samples = 0
-            self._validation_dl = self.build_pretraining_data_loader(self._validation_ds, consumed_samples)
+            self._validation_dl = self.build_pretraining_data_loader(self._validation_ds, consumed_samples, num_workers=0)
```
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:893]: Should `num_workers=0` be hard coded here?

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py [hunk 4]
```diff
@@ -823,12 +890,12 @@ def on_pretrain_routine_start(self) -> None:
...
 
     def setup_test_data(self, cfg):
         if hasattr(self, '_test_ds'):
             consumed_samples = 0
-            self._test_dl = self.build_pretraining_data_loader(self._test_ds, consumed_samples)
+            self._test_dl = self.build_pretraining_data_loader(self._test_ds, consumed_samples, num_workers=0)
```
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:898]: Should `num_workers=0` be hard coded here?

Comment (MaximumEntropy) [reply nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py]: Good point. Reverted.

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py (hunk 3 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py]: Reverted.

**Review by khcs — APPROVED** [review]
ready to be merged.

**Pull request merged**